### PR TITLE
DEV: Move logic for rate limiting user second factor to one place

### PR DESCRIPTION
**Pull request by martin-brennan**
This moves all the rate limiting for user second factor (based on `params[:second_factor_token]` existing) to the one place, which rate limits by IP and also by username if a user is found.

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -542,6 +542,16 @@ def handle_permalink(path)
     end
   end
 
+  def rate_limit_second_factor!(user)
+    return if params[:second_factor_token].blank?
+
+    RateLimiter.new(nil, "second-factor-min-#{request.remote_ip}", 3, 1.minute).performed!
+
+    if user
+      RateLimiter.new(nil, "second-factor-min-#{user.username}", 3, 1.minute).performed!
```
Comment (martin-brennan) [inline app/controllers/application_controller.rb:551]: Not sure if we want to do `RateLimiter.new(nil, "second-factor-min-anonymous", 3, 1.minute).performed!` if the user does not exist?

**Comment by SamSaffron** [comment]
I would like to take this a bit further, are there other places we check second factor? seeing this is user_email_controller, users_controller as well. We would like to centralize this new rate limit maybe chuck it all the way into `user.authenticate_second_factor` I don't know.

**Comment by martin-brennan** [comment]
@SamSaffron the places I changed in this PR are the only places that we appear to check second factor. I did consider moving it to `user.authenticate_second_factor`, but then that requires the existence of the user, which means we cannot do just the IP rate limit on the second factor endpoint even when a user does not exist.

**Comment by martin-brennan** [comment]
Hmm wait a minute I found another, I do not know how I missed it, on password_reset_update, will patch that

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (SamSaffron) [reply app/controllers/application_controller.rb:551]: I think this is ok, but we are going to hit this multiple times in a few workflows lets double it from 3 a minute to 6 a minute

**Comment by SamSaffron** [comment]
This is looking good, but I think a tiny bit more test coverage is a good idea.

**Comment by martin-brennan** [comment]
@SamSaffron added more coverage now :rocket:

**Comment by SamSaffron** [comment]
approving this for merge backport, just remember to up the rate limit from 3 per minute to 6.

**Review by SamSaffron — APPROVED** [review]

**Comment by martin-brennan** [comment]
@SamSaffron OK upped the rate limit, will merge and backport tomorrow

File: app/controllers/session_controller.rb [hunk 2]
```diff
@@ -277,7 +276,10 @@ def create
 
     return invalid_credentials if params[:password].length > User.max_password_length
 
-    if user = User.find_by_username_or_email(normalized_login_param)
+    user = User.find_by_username_or_email(normalized_login_param)
+    rate_limit_second_factor!(user)
```
Comment (ZogStriP) [inline app/controllers/session_controller.rb:280]: What is `user` is `nil`?

**Pull request merged**